modules/distributor/forwarder/manager.go

```diff
@@ -0,0 +1,370 @@
+	ql, found := m.tenantToQueueList[tenantID]
+	if !found {
+		_ = level.Warn(m.logger).Log("msg", "queue list not found", "tenantID", tenantID)
+		return
+	}
+
+	if err := ql.shutdown(context.TODO()); err != nil {
+		_ = level.Warn(m.logger).Log("msg", "failed to shutdown queue list", "tenantID", tenantID)
+	}
+
+	delete(m.tenantToQueueList, tenantID)
+}
+
+// updateQueueListForTenantUnderLock should only be called under tenantToQueueListMu write lock.
+func (m *Manager) updateQueueListForTenantUnderLock(tenantID string, forwarderNames []string) {
+	queueList, found := m.tenantToQueueList[tenantID]
+	if !found {
+		queueList = newQueueList(m.logger, m.reg, tenantID)
+		m.tenantToQueueList[tenantID] = queueList
+	}
+
+	queueList.update(forwarderNames, m.forwarderNameToForwarder)
+}
+
+func (m *Manager) getQueueList(tenantID string) (*queueList, bool) {
+	m.tenantToQueueListMu.RLock()
+	defer m.tenantToQueueListMu.RUnlock()
+
+	queueList, found := m.tenantToQueueList[tenantID]
+
+	return queueList, found
+}
+
+func (m *Manager) shutdown() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	m.tenantToQueueListMu.Lock()
+	defer m.tenantToQueueListMu.Unlock()
+
+	var errs []error
+	for tenantID, ql := range m.tenantToQueueList {
+		if err := ql.shutdown(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("failed to shutdown queuelist for tenantID=%s: %w", tenantID, err))
+		}
+
+		delete(m.tenantToQueueList, tenantID)
+	}
+
+	for forwarderName, forwarder := range m.forwarderNameToForwarder {
+		if err := forwarder.Shutdown(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("failed to shutdown forwarder with name=%s: %w", forwarderName, err))
+		}
+
+		delete(m.forwarderNameToForwarder, forwarderName)
+	}
+
+	return multierr.Combine(errs...)
+}
+
+type queueListDiff[T any] struct {
+	added   []T
+	removed []T
+}
+
+type queueList struct {
+	logger               log.Logger
+	reg                  prometheus.Registerer
+	tenantID             string
+	forwarderNameToQueue map[string]*queue.Queue[tempopb.Trace]
+	mu                   *sync.RWMutex
+}
+
+func newQueueList(logger log.Logger, reg prometheus.Registerer, tenantID string) *queueList {
+	return &queueList{
+		logger:               logger,
+		reg:                  reg,
+		tenantID:             tenantID,
+		forwarderNameToQueue: make(map[string]*queue.Queue[tempopb.Trace]),
+		mu:                   &sync.RWMutex{},
+	}
+}
+
+func (l *queueList) copy() List {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	result := make(List, 0, len(l.forwarderNameToQueue))
+	for _, q := range l.forwarderNameToQueue {
+		result = append(result, &queueAdapter{queue: q})
+	}
+
+	return result
+}
+
+func (l *queueList) update(forwarderNames []string, forwarderNameToForwarder map[string]Forwarder) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	oldForwarderNames := make([]string, 0, len(l.forwarderNameToQueue))
```

Comment:
similar to above there is likely a more straightforward way to write this.

Reply:
Yeah, similar comment here.